src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -93,6 +105,9 @@ static async ValueTask<QuicConnection> StartConnectAsync(QuicClientConnectionOpt
     private readonly ValueTaskSource _connectedTcs = new ValueTaskSource();
     private readonly ValueTaskSource _shutdownTcs = new ValueTaskSource();
 
+    private readonly CancellationTokenSource _shutdownCancellationTokenSource = new CancellationTokenSource();
```

Comment:
I think we might need to dispose `_shutdownCancellationTokenSource`?
Also, the naming confuses me... It implies the cancellation of the connection's shutdown, but it's not, it's for cancelling the handshake in the Listener, right? (and it also means it's not needed on the client side?)

Reply:
It is used only for listener to cancel the connection options callback if peer closes the connection before the handshake. I did not find a way to put the CTS outside QuicConnection.

I will try to come up with a better name

Reply:
What do we do in SslStream case? Do we also react and cancel user callback when the client kills handshake in the meantime?

Reply:
SslStream has only synchronous callback

Reply:
With sslstream we don't check for client-side activity while invoking the callback. We would have to always have a pending read on the inner stream and check for errors there. SslStream passes in the same token as to AuthenticateAsClientAsync.

Reply:
Renamed to `ConnectionShutdownToken` and added a comment